strings/base_collections_base.h

```diff
@@ -189,64 +286,85 @@ WINRT_EXPORT namespace winrt
 
         void SetAt(uint32_t const index, T const& value)
         {
-            if (index >= static_cast<D const&>(*this).get_container().size())
+            static_cast<D&>(*this).perform_exclusive([&]
             {
-                throw hresult_out_of_bounds();
-            }
+                if (index >= static_cast<D const&>(*this).get_container().size())
+                {
+                    throw hresult_out_of_bounds();
+                }
 
-            this->increment_version();
-            static_cast<D&>(*this).get_container()[index] = static_cast<D const&>(*this).wrap_value(value);
+                this->increment_version();
+                static_cast<D&>(*this).get_container()[index] = static_cast<D const&>(*this).wrap_value(value);
+            });
         }
 
         void InsertAt(uint32_t const index, T const& value)
         {
-            if (index > static_cast<D const&>(*this).get_container().size())
+            static_cast<D&>(*this).perform_exclusive([&]
             {
-                throw hresult_out_of_bounds();
-            }
+                if (index > static_cast<D const&>(*this).get_container().size())
+                {
+                    throw hresult_out_of_bounds();
+                }
 
-            this->increment_version();
-            static_cast<D&>(*this).get_container().insert(static_cast<D const&>(*this).get_container().begin() + index, static_cast<D const&>(*this).wrap_value(value));
+                this->increment_version();
+                static_cast<D&>(*this).get_container().insert(static_cast<D const&>(*this).get_container().begin() + index, static_cast<D const&>(*this).wrap_value(value));
+            });
         }
 
         void RemoveAt(uint32_t const index)
         {
-            if (index >= static_cast<D const&>(*this).get_container().size())
+            static_cast<D&>(*this).perform_exclusive([&]
             {
-                throw hresult_out_of_bounds();
-            }
+                if (index >= static_cast<D const&>(*this).get_container().size())
+                {
+                    throw hresult_out_of_bounds();
+                }
 
-            this->increment_version();
-            static_cast<D&>(*this).get_container().erase(static_cast<D const&>(*this).get_container().begin() + index);
+                this->increment_version();
+                static_cast<D&>(*this).get_container().erase(static_cast<D const&>(*this).get_container().begin() + index);
```

Comment:
This also destructs the old item under the lock. Old value needs to be moved to an object that is created outside the lock, so that the destruction happens outside the lock.